crates/sparrow-runtime/src/read/parquet_file.rs

```diff
@@ -68,58 +77,78 @@ impl ParquetFile {
 
         Ok(Self {
             object_store,
-            path,
-            metadata,
+            object_meta,
+            parquet_metadata: metadata,
             schema,
         })
     }
 
+    pub fn num_rows(&self) -> usize {
+        self.parquet_metadata.file_metadata().num_rows() as usize
+    }
+
     pub async fn read_stream(
-        self,
+        &self,
+        batch_size: Option<usize>,
         projection: Option<Vec<usize>>,
     ) -> error_stack::Result<BoxStream<'static, error_stack::Result<RecordBatch, Error>>, Error>
     {
-        let path = self.path.clone();
-        let metadata = self.metadata.clone();
+        let reader = AsyncParquetObjectReader {
+            object_store: self.object_store.clone(),
+            object_meta: self.object_meta.clone(),
+            parquet_metadata: self.parquet_metadata.clone(),
+        };
 
-        let mut batch_stream = ParquetRecordBatchStreamBuilder::new(self)
+        let mut batch_stream = ParquetRecordBatchStreamBuilder::new(reader)
             .await
             .into_report()
             .change_context(Error::ReadingParquetFile)
-            .attach_printable_lazy(|| path.clone())?;
+            .attach_printable_lazy(|| self.object_meta.location.clone())?;
 
-        batch_stream = batch_stream.with_batch_size(BATCH_SIZE_ROWS);
+        batch_stream = batch_stream.with_batch_size(batch_size.unwrap_or(BATCH_SIZE_ROWS));
```

Comment:
Can you explain how this works?

Reply:
If the `batch_size` is `None` then the default value (`BATCH_SIZE_ROWS` is used). If it is `Some(size)` then it is `size`.